File: merge-two-sorted-list/merge-two-sorted-list.go
package main

import "fmt"

type ListNode struct {
	Val  int
	Next *ListNode
}

func main() {
	list11 := ListNode{2, nil}
	list12 := ListNode{4, nil}
	list13 := ListNode{3, nil}
	list11.Next = &list12
	list12.Next = &list13
	list21 := ListNode{5, nil}
	list22 := ListNode{6, nil}
	list23 := ListNode{4, nil}
	list21.Next = &list22
	list22.Next = &list23
	list31 := mergeTwoLists(&list11, &list21)
	for list31 != nil {
		fmt.Printf("%d ", list31.Val)
		list31 = list31.Next
	}
}

func mergeTwoLists(list1 *ListNode, list2 *ListNode) *ListNode {
	result := ListNode{}
	current := &result
	for list1 != nil || list2 != nil {
		if list1 != nil && list2 != nil {
			if list1.Val < list2.Val {
				current.Next = list1
				list1 = list1.Next
			} else {
				current.Next = list2
				list2 = list2.Next
			}
		} else if list2 != nil {
			current.Next = list2
			list2 = list2.Next
		} else {
			current.Next = list1
			list1 = list1.Next
		}
		current = current.Next
	}
	return result.Next
}

File: hackerrank-java/src/com/company/MergeTwoSortedList.java
package com.company;

import java.util.List;

public class MergeTwoSortedList {

    public static ListNode mergeTwoLists(ListNode list1, ListNode list2) {

        ListNode result = new ListNode();
        ListNode current = result;
        while (list1 != null || list2 != null) {
            if (list1 != null && list2 != null){
                if (list1.val < list2.val) {
                    current.next = list1;
                    list1 = list1.next;
                } else {
                    current.next = list2;
                    list2 = list2.next;
                }
            } else if(list2 != null) {
                current.next = list2;
                list2 = list2.next;
            } else {
                current.next = list1;
                list1 = list1.next;
            }
            current = current.next;
        }
        return result.next;
    }


    public static ListNode mergeTwoListsRecursive(ListNode list1, ListNode list2) {

        ListNode current = new ListNode();
        ListNode head = mergeTwoListsRecursiveInternal(list1, list2, current);
        return current.next;
    }

    // 1 3 5 7 9
    // 2 4 6 8 10

    // 1  2
    // dummy ->
    //
    //

    private static ListNode mergeTwoListsRecursiveInternal(ListNode list1, ListNode list2, ListNode current) {

        if (list1 != null && list2 != null){
            if (list1.val < list2.val) {
                current.next = new ListNode(list1.val);
                return mergeTwoListsRecursiveInternal(list1.next, list2, current.next);
            } else {
                current.next = new ListNode(list2.val);
                return mergeTwoListsRecursiveInternal(list1, list2.next, current.next);
            }
        }

        if ( list1 != null ){
            current.next = new ListNode(list1.val);
            return mergeTwoListsRecursiveInternal(list1.next, null, current.next);
        }
        if ( list2 != null ){
            current.next = new ListNode(list2.val);
            return mergeTwoListsRecursiveInternal(null, list2.next, current.next);
        }
        return null;
    }


    public static void main(String[] args) {
        ListNode list11 = new ListNode(1) ;
        ListNode list12 = new ListNode(3) ;
        ListNode list13 = new ListNode(5) ;
        list11.next = list12; list12.next = list13;
        ListNode list21 = new ListNode(2) ;
        ListNode list22 = new ListNode(4) ;
        ListNode list23 = new ListNode(6) ;
        list21.next = list22; list22.next = list23;
        ListNode list31 = mergeTwoListsRecursive(list11, list21);
        while(list31 != null) {
            System.out.print(list31.val+  " ");
            list31 = list31.next;
        }
    }
}
